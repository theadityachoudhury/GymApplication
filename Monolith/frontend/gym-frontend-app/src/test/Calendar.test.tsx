/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable react-hooks/rules-of-hooks */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi, beforeAll, afterAll } from 'vitest';
import CalendarContainer from '../components/common/form/Calendar';

// Mock react-datepicker completely
vi.mock('react-datepicker', () => {
	return {
		__esModule: true,
		default: (props: any) => {
			// Get the fixed today date for testing
			// const today = new Date('2023-06-15');

			// Call the onMonthChange prop to simulate month changes
			React.useEffect(() => {
				if (props.onMonthChange) {
					const mockMonthChangeHandler = () => {
						// This is just to register that the handler exists
					};
					mockMonthChangeHandler();
				}
			}, [props.onMonthChange]);

			// Call the onChange prop to simulate date selection
			React.useEffect(() => {
				if (props.onChange) {
					const mockChangeHandler = () => {
						// This is just to register that the handler exists
					};
					mockChangeHandler();
				}
			}, [props.onChange]);

			// Generate test dates with proper past/future status
			const testDates = [
				{ day: 10, date: new Date('2023-06-10') }, // Past date
				{ day: 15, date: new Date('2023-06-15') }, // Today
				{ day: 20, date: new Date('2023-06-20') }, // Future date (selected)
				{ day: 25, date: new Date('2023-06-25') }, // Future date
			];

			// Render a simplified version that just shows the important parts
			return (
				<div data-testid="datepicker-mock">
					{/* Render the custom header */}
					{props.renderCustomHeader &&
						props.renderCustomHeader({
							date: props.selected || new Date(),
							decreaseMonth: vi.fn(),
							increaseMonth: vi.fn(),
							prevMonthButtonDisabled: false,
							nextMonthButtonDisabled: false,
						})}

					{/* Mock day names */}
					<div className="react-datepicker__day-names">
						<div>SUN</div>
						<div>MON</div>
						<div>TUE</div>
						<div>WED</div>
						<div>THU</div>
						<div>FRI</div>
						<div>SAT</div>
					</div>

					{/* Mock test dates */}
					<div className="react-datepicker__month">
						{testDates.map(({ day, date }) => (
							<div key={day} data-testid={`day-${day}`}>
								{props.renderDayContents &&
									props.renderDayContents(day, date)}
							</div>
						))}
					</div>
				</div>
			);
		},
	};
});

// Mock the Lucide React icons
vi.mock('lucide-react', () => ({
	ChevronLeft: () => <div data-testid="chevron-left-icon">←</div>,
	ChevronRight: () => <div data-testid="chevron-right-icon">→</div>,
}));

describe('CalendarContainer Component', () => {
	// Fixed date for consistent testing
	const mockToday = new Date('2023-06-15');
	const mockSelectedDate = new Date('2023-06-20');

	// Mock the current date
	beforeAll(() => {
		vi.useFakeTimers();
		vi.setSystemTime(mockToday);
	});

	afterAll(() => {
		vi.useRealTimers();
	});

	// Basic rendering tests
	it('renders the calendar component', () => {
		const handleDateSelect = vi.fn();
		render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		// Check if the datepicker mock is rendered
		expect(screen.getByTestId('datepicker-mock')).toBeInTheDocument();
	});

	it('displays the correct month and year in the header', () => {
		const handleDateSelect = vi.fn();
		render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		expect(screen.getByText('June 2023')).toBeInTheDocument();
	});

	it('renders navigation buttons', () => {
		const handleDateSelect = vi.fn();
		render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		expect(screen.getByTestId('chevron-left-icon')).toBeInTheDocument();
		expect(screen.getByTestId('chevron-right-icon')).toBeInTheDocument();
	});

	it('renders weekday headers', () => {
		const handleDateSelect = vi.fn();
		render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		// Check for abbreviated weekday names
		expect(screen.getByText('SUN')).toBeInTheDocument();
		expect(screen.getByText('MON')).toBeInTheDocument();
		expect(screen.getByText('TUE')).toBeInTheDocument();
		expect(screen.getByText('WED')).toBeInTheDocument();
		expect(screen.getByText('THU')).toBeInTheDocument();
		expect(screen.getByText('FRI')).toBeInTheDocument();
		expect(screen.getByText('SAT')).toBeInTheDocument();
	});

	// Test day rendering
	it('highlights the selected date', () => {
		const handleDateSelect = vi.fn();
		render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		// Find the selected date (20th)
		const selectedDateElement = screen.getByTestId('day-20').firstChild;

		// Check if it has the selected styling
		expect(selectedDateElement).toHaveClass('bg-green-100');
		expect(selectedDateElement).toHaveClass('border-2');
		expect(selectedDateElement).toHaveClass('border-green-300');
	});

	it('shows past dates as disabled', () => {
		const handleDateSelect = vi.fn();
		render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		// Find a past date (e.g., the 10th, which is before our mock today of the 15th)
		const pastDateElement = screen.getByTestId('day-10').firstChild;

		// Check if it has the disabled styling
		expect(pastDateElement).toHaveClass('text-gray-300');
	});

	it('shows future dates as enabled', () => {
		const handleDateSelect = vi.fn();
		render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		// Find a future date (e.g., the 25th, which is after our mock today of the 15th)
		const futureDateElement = screen.getByTestId('day-25').firstChild;

		// Check if it has the enabled styling
		expect(futureDateElement).toHaveClass('text-gray-700');
		expect(futureDateElement).toHaveClass('hover:bg-gray-100');
	});

	// Styling tests
	it('applies custom styling to the calendar', () => {
		const handleDateSelect = vi.fn();
		const { container } = render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		// Check for the presence of custom styles
		const styleElement = container.querySelector('style');
		expect(styleElement).toBeInTheDocument();
		expect(styleElement?.textContent).toContain("font-family: 'Lexend'");
	});

	// Snapshot test
	it('matches snapshot', () => {
		const handleDateSelect = vi.fn();
		const { container } = render(
			<CalendarContainer
				selectedDate={mockSelectedDate}
				onDateSelect={handleDateSelect}
			/>
		);

		expect(container).toMatchSnapshot();
	});
});
