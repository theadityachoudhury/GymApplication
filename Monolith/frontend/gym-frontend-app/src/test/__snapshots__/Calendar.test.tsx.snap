// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalendarContainer Component > matches snapshot 1`] = `
<div>
  <div
    class="w-full font-lexend"
  >
    <link
      href="https://fonts.googleapis.com/css2?family=Lexend:wght@300;400;500;600;700&display=swap"
      rel="stylesheet"
    />
    <div
      data-testid="datepicker-mock"
    >
      <div>
        <div
          class="flex justify-between items-center p-4 font-lexend w-full"
        >
          <button
            class="p-1 rounded-full hover:bg-gray-100 text-gray-500"
          >
            <div
              data-testid="chevron-left-icon"
            >
              ←
            </div>
          </button>
          <div
            class="text-xl font-lexend"
          >
            June 2023
          </div>
          <button
            class="p-1 rounded-full hover:bg-gray-100 text-gray-400"
          >
            <div
              data-testid="chevron-right-icon"
            >
              →
            </div>
          </button>
        </div>
        <hr
          class="border-t border-gray-200 mx-4"
        />
      </div>
      <div
        class="react-datepicker__day-names"
      >
        <div>
          SUN
        </div>
        <div>
          MON
        </div>
        <div>
          TUE
        </div>
        <div>
          WED
        </div>
        <div>
          THU
        </div>
        <div>
          FRI
        </div>
        <div>
          SAT
        </div>
      </div>
      <div
        class="react-datepicker__month"
      >
        <div
          data-testid="day-10"
        >
          <div
            class="flex flex-col items-center"
          >
            <div
              class="
          flex flex-col items-center justify-center h-10 w-10 rounded-full mx-auto font-lexend
          text-gray-300
          
          
          
        "
            >
              10
            </div>
          </div>
        </div>
        <div
          data-testid="day-15"
        >
          <div
            class="flex flex-col items-center"
          >
            <div
              class="
          flex flex-col items-center justify-center h-10 w-10 rounded-full mx-auto font-lexend
          text-gray-700
          
          hover:bg-gray-100 hover:rounded-full
          
        "
            >
              15
            </div>
          </div>
        </div>
        <div
          data-testid="day-20"
        >
          <div
            class="flex flex-col items-center"
          >
            <div
              class="
          flex flex-col items-center justify-center h-10 w-10 rounded-full mx-auto font-lexend
          text-gray-700
          bg-green-100 border-2 border-green-300
          
          
        "
            >
              20
            </div>
          </div>
        </div>
        <div
          data-testid="day-25"
        >
          <div
            class="flex flex-col items-center"
          >
            <div
              class="
          flex flex-col items-center justify-center h-10 w-10 rounded-full mx-auto font-lexend
          text-gray-700
          
          hover:bg-gray-100 hover:rounded-full
          
        "
            >
              25
            </div>
          </div>
        </div>
      </div>
    </div>
    <style>
      
        @import url('https://fonts.googleapis.com/css2?family=Lexend:wght@300;400;500;600;700&display=swap');

        .font-lexend {
          font-family: 'Lexend', sans-serif;
        }

        .react-datepicker {
          border: none;
          width: 100%;
          font-family: 'Lexend', sans-serif;
        }

        .react-datepicker * {
          font-family: 'Lexend', sans-serif;
        }

        .react-datepicker__month-container {
          width: 100%;
        }

        .react-datepicker__month {
          margin: 0;
          padding: 0 1rem 1rem;
        }

        .react-datepicker__week {
          display: flex;
          justify-content: space-between;
        }

        .react-datepicker__day {
          display: flex;
          align-items: center;
          justify-content: center;
          margin: 0;
          padding: 0;
          width: 40px;
          height: 40px;
        }

        .react-datepicker__header {
          background-color: white;
          border-bottom: none;
        }

        .react-datepicker__day-names {
          display: flex;
          justify-content: space-between;
          padding: 0 1rem;
          margin-bottom: 0.5rem;
        }

        .react-datepicker__day-name {
          width: 40px;
          height: 40px;
          display: flex;
          align-items: center;
          justify-content: center;
          margin: 0;
          color: #666;
          font-weight: 600;
          font-size: 0.75rem;
        }

        .react-datepicker__day--selected,
        .react-datepicker__day--keyboard-selected,
        .react-datepicker__day--in-selecting-range,
        .react-datepicker__day--in-range {
          background-color: transparent !important;
          color: black !important;
        }

        .react-datepicker__day:focus,
        .react-datepicker__day--selected:focus,
        .react-datepicker__day--keyboard-selected:focus {
          outline: none !important;
        }

        .react-datepicker__day:hover,
        .react-datepicker__day--selected:hover,
        .react-datepicker__day--keyboard-selected:hover {
          border-radius: 50% !important;
          background-color: #f3f4f6 !important;
        }
        
        .react-datepicker__day--excluded {
          cursor: not-allowed;
          color: #ccc !important;
        }
      
    </style>
  </div>
</div>
`;
